ingestion-beam/src/main/java/com/mozilla/telemetry/heka/HekaReader.java

```diff
@@ -0,0 +1,145 @@
+        if (f.getValueType() == Heka.Field.ValueType.STRING && target.isObject()) {
+          String value = f.getValueString(0);
+          if (value.charAt(0) == '{') {
+            target.set(lastKey, Json.readObjectNode(value.getBytes(StandardCharsets.UTF_8)));
+          } else if (value.charAt(0) == '[') {
+            target.set(lastKey, Json.readArrayNode(value.getBytes(StandardCharsets.UTF_8)));
+          } else {
+            target.put(lastKey, value);
+          }
+        } else if (f.getValueType() == Heka.Field.ValueType.BOOL) {
+          target.put(lastKey, f.getValueBool(0));
+        } else if (f.getValueType() == ValueType.INTEGER) {
+          target.put(lastKey, f.getValueInteger(0));
+        } else if (f.getValueType() == Heka.Field.ValueType.DOUBLE) {
+          target.put(lastKey, f.getValueDouble(0));
+        }
+        // FIXME: do we need to handle byte fields?
+        // see code in moztelemetry and
+        // https://bugzilla.mozilla.org/show_bug.cgi?id=1339421
+      }
+    }
+
+    Map<String, String> attributes = new HashMap<>();
+    attributes.put(Attribute.DOCUMENT_NAMESPACE, "telemetry");
+    attributes.put(Attribute.HOST, message.getHostname());
+    Optional.ofNullable(payload.remove("meta")).ifPresent(meta -> {
+      Optional.ofNullable(meta.path("Date").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DATE, s));
+      Optional.ofNullable(meta.path("docType").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DOCUMENT_TYPE, s));
+      Optional.ofNullable(meta.path("appBuildId").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_BUILD_ID, s));
+      Optional.ofNullable(meta.path("appName").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_NAME, s));
+      Optional.ofNullable(meta.path("appUpdateChannel").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_UPDATE_CHANNEL, s));
+      Optional.ofNullable(meta.path("appVersion").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_VERSION, s));
+      Optional.ofNullable(meta.path("documentId").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DOCUMENT_ID, s));
+      Optional.ofNullable(meta.path("sourceVersion").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DOCUMENT_VERSION, s));
+      Optional.ofNullable(meta.path("geoCity").textValue())
+          .ifPresent(s -> attributes.put(Attribute.GEO_CITY, s));
+      Optional.ofNullable(meta.path("geoCountry").textValue())
+          .ifPresent(s -> attributes.put(Attribute.GEO_COUNTRY, s));
+      Optional.ofNullable(meta.path("geoSubdivision1").textValue())
+          .ifPresent(s -> attributes.put(Attribute.GEO_SUBDIVISION1, s));
+      Optional.ofNullable(meta.path("geoSubdivision2").textValue())
+          .ifPresent(s -> attributes.put(Attribute.GEO_SUBDIVISION2, s));
+      // TODO: Do heka messages contain parsed user agent info? DNT? Method? Protocol?
```

Comment:
@mreid-moz Do heka messages contain parsed user agent info? Other headers?

Reply:
I don't think we parse/store user agent info in the telemetry data on AWS.

cc @trink can you confirm?

Reply:
We do not store/parse user agent info for telemetry data in AWS (we do for structured). We do store these headers: https://github.com/mozilla-services/lua_sandbox_extensions/blob/master/moz_telemetry/io_modules/decoders/moz_ingest/telemetry.lua#L321-L324.